packages/plugin-pt15/src/algorithm.ts

```diff
@@ -0,0 +1,214 @@
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+  PositionStorage,
+]
+
+export interface PT15IndexStore extends AnyIndexStore {
+  indexes: Record<string, Tree>
+  vectorIndexes: Record<string, VectorIndex>
+  searchableProperties: string[]
+  searchablePropertiesWithTypes: Record<string, SearchableType>
+}
+
+function create_obj() {
+  // object with empty prototype to cheap objects
+  return Object.create(null)
+}
+
+export function recursiveCreate<T extends AnyOrama>(indexDatastore: PT15IndexStore, schema: T['schema'], prefix: string) {
+  for (const [prop, type] of Object.entries<SearchableType>(schema)) {
+    const path = `${prefix}${prefix ? '.' : ''}${prop}`
+
+    if (typeof type === 'object' && !Array.isArray(type)) {
+      // Nested
+      recursiveCreate(indexDatastore, type, path)
+      continue
+    }
+
+    if (isVectorType(type)) {
+      indexDatastore.searchableProperties.push(path)
+      indexDatastore.searchablePropertiesWithTypes[path] = type
+      indexDatastore.vectorIndexes[path] = {
+        size: getVectorSize(type),
+        vectors: {}
+      }
+    } else {
+      const isArray = /\[/.test(type as string)
+      switch (type) {
+        case 'boolean':
+        case 'boolean[]':
+          indexDatastore.indexes[path] = { type: 'Bool', node: new bool.BoolNode(), isArray }
+          break
+        case 'number':
+        case 'number[]':
+          indexDatastore.indexes[path] = { type: 'AVL', node: new avl.AVLTree<number, InternalDocumentID[]>(0, []), isArray }
+          break
+        case 'string':
+        case 'string[]':
+          indexDatastore.indexes[path] = { type: 'Position', node: [
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+            create_obj(),
+          ], isArray }
+          break
+        case 'enum':
+        case 'enum[]':
+          indexDatastore.indexes[path] = { type: 'Flat', node: new flat.FlatTree(), isArray }
+          break
+        case 'geopoint':
+          indexDatastore.indexes[path] = { type: 'BKD', node: new bkd.BKDTree(), isArray }
+          break
+        default:
+          throw new Error('INVALID_SCHEMA_TYPE: ' + path)
+      }
+
+      indexDatastore.searchableProperties.push(path)
+      indexDatastore.searchablePropertiesWithTypes[path] = type
+    }
+  }
+}
+
+export function insertString(
+  value: string,
+  positionsStorage: PositionsStorage,
+  prop: string,
+  internalId: InternalDocumentID,
+  language: string | undefined,
```

Comment:
I think I started exposing a `Optional` type which is basically `type Optional<T = unknown> = T | undefined`